compile error `vfs_parse_fs_param_source` with kernel 5.10.x

there might be a compile error with kernel 5.10.x

```text
  CC [M]  /build/source/module/os/linux/zfs/zpl_file_range.o
  CC [M]  /build/source/module/os/linux/zfs/zpl_inode.o
  CC [M]  /build/source/module/os/linux/zfs/zpl_super.o
  CC [M]  /build/source/module/os/linux/zfs/zpl_xattr.o
  CC [M]  /build/source/module/os/linux/zfs/zvol_os.o
/build/source/module/os/linux/zfs/zpl_super.c: In function 'zpl_parse_param':
/build/source/module/os/linux/zfs/zpl_super.c:618:19: error: implicit declaration of function 'vfs_parse_fs_param_source'; did you mean 'vfs_parse_fs_param'? [8;;https://gcc.gnu.org/onlinedocs/gcc-15.2.0/gcc/Warning-Options.html#index-Wimplicit-function-declaration�-Wimplicit-function-declaration8;;�]
  618 |         int opt = vfs_parse_fs_param_source(fc, param);
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~
      |                   vfs_parse_fs_param
make[5]: *** [/nix/store/cw11bnbbd4id2z4l91v473fi9v15c4qd-linux-5.10.255-dev/lib/modules/5.10.255/source/scripts/Makefile.build:286: /build/source/module/os/linux/zfs/zpl_super.o] Error 1
make[5]: *** Waiting for unfinished jobs....
make[4]: *** [/nix/store/cw11bnbbd4id2z4l91v473fi9v15c4qd-linux-5.10.255-dev/lib/modules/5.10.255/source/Makefile:1853: /build/source/module] Error 2
make[3]: *** [/nix/store/cw11bnbbd4id2z4l91v473fi9v15c4qd-linux-5.10.255-dev/lib/modules/5.10.255/source/Makefile:192: __sub-make] Error 2
make[3]: Leaving directory '/nix/store/cw11bnbbd4id2z4l91v473fi9v15c4qd-linux-5.10.255-dev/lib/modules/5.10.255/build'
make[2]: *** [Makefile:57: modules-Linux] Error 2
make[2]: Leaving directory '/build/source/module'
make[1]: *** [Makefile:12775: all-recursive] Error 1
make[1]: Leaving directory '/build/source'
make: *** [Makefile:4954: all] Error 2
```

this was observed during a build of OpenZFS 2.4.2 on Linux 5.10.255 (both x86_64-linux and aarch64-linux)

https://github.com/matdibu/nixpkgs-review-gha/actions/runs/25782026442

5.10.x does not have the `vfs_parse_fs_param_source` function, only `vfs_parse_fs_param`

https://github.com/torvalds/linux/blob/2c85ebc57b3e1817b6ce1a6b703928e113a90442/include/linux/fs_context.h#L136

it was added in 5.14

https://github.com/torvalds/linux/blob/7d2a07b769330c34b4deabeed939325c77a7ec2f/include/linux/fs_context.h#L142

_Originally posted by @matdibu in https://github.com/openzfs/zfs/issues/18377#issuecomment-4438835599_

--- a/module/os/linux/zfs/zpl_super.c
+++ b/module/os/linux/zfs/zpl_super.c
@@ -550,10 +550,11 @@ zpl_prune_sb(uint64_t nr_to_scan, void *arg)
  *
  * Finally, all filesystems get automatic handling for the 'source' option,
  * that is, the "name" of the filesystem (the first column of df(1)'s output).
- * However, this only happens if the handler does not otherwise handle
- * the 'source' option. Since we handle _all_ options because of 'sloppy', we
- * deal with this explicitly by calling into the kernel's helper for this,
- * vfs_parse_fs_param_source(), which sets up fc->source.
+ * However, this only happens if the handler does not otherwise handle the
+ * 'source' option. Since we handle _all_ options because of 'sloppy', we have
+ * ot handle it ourselves. Normally we would call vfs_parse_fs_param_source()
+ * to deal with this, but that didn't appear until 5.14, and it's small enough
+ * that we can just handle it ourselves.
  *
  *	source
  *
@@ -565,6 +566,7 @@ zpl_prune_sb(uint64_t nr_to_scan, void *arg)
  */
 
 enum {
+	Opt_source,
 	Opt_exec, Opt_suid, Opt_dev,
 	Opt_atime, Opt_relatime, Opt_strictatime,
 	Opt_saxattr, Opt_dirxattr, Opt_noxattr,
@@ -574,6 +576,8 @@ enum {
 };
 
 static const struct fs_parameter_spec zpl_param_spec[] = {
+	fsparam_string("source",	Opt_source),
+
 	fsparam_flag_no("exec",		Opt_exec),
 	fsparam_flag_no("suid",		Opt_suid),
 	fsparam_flag_no("dev",		Opt_dev),
@@ -614,13 +618,8 @@ zpl_parse_param(struct fs_context *fc, struct fs_parameter *param)
 {
 	vfs_t *vfs = fc->fs_private;
 
-	/* Handle 'source' explicitly so we don't trip on it as an unknown. */
-	int opt = vfs_parse_fs_param_source(fc, param);
-	if (opt != -ENOPARAM)
-		return (opt);
-
 	struct fs_parse_result result;
-	opt = fs_parse(fc, zpl_param_spec, param, &result);
+	int opt = fs_parse(fc, zpl_param_spec, param, &result);
 	if (opt == -ENOPARAM) {
 		/*
 		 * Convert unknowns to warnings, to work around the whole
@@ -632,6 +631,16 @@ zpl_parse_param(struct fs_context *fc, struct fs_parameter *param)
 		return (opt);
 
 	switch (opt) {
+	case Opt_source:
+		if (fc->source != NULL) {
+			cmn_err(CE_NOTE,
+			    "ZFS: multiple 'source' options not supported");
+			return (-SET_ERROR(EINVAL));
+		}
+		fc->source = param->string;
+		param->string = NULL;
+		break;
+
 	case Opt_exec:
 		vfs->vfs_exec = !result.negated;
 		vfs->vfs_do_exec = B_TRUE;

